content/2_2_redis.tex

```diff
@@ -0,0 +1,362 @@
+\texttt{HMGET}). \parencite{redis:dataTypesIntroduction, redis:commands,
+redis:dataTypes}
+
+\subsubsection*{Sorted Sets}
+Similar to sets, sorted sets are non repeating collections of strings ordered by
+a non-unique score (smallest to greatest).  Members within a sorted set can be
+added, removed and returned (\texttt{ZADD}, \texttt{ZPOPMIN}, \texttt{ZPOPMAX}).
+Moreover it is possible to return members with a certain score or at a certain
+position/index within the sorted set. Also, scores can be increased and thereby
+updated.
+
+Thus sorted sets can be used to keep track of any kind of ranking like a
+competition. In this case, scores can be initially inserted and later updated using
+\texttt{ZADD}.  Due to operations like \texttt{ZRANGE} or \texttt{ZRANK}, it is
+possible to receive the top or bottom half of the ranking, or receive the rank
+of a specific member. \parencite{redis:dataTypesIntroduction, redis:commands,
+redis:dataTypes}
+
+\subsection{Multi-node setups / Redis Cluster} \label{redis:multiNodeSetup}
+Originally, Redis only supported single-node and non-clustered setups. In
+combination with its mostly single-threaded architecture, this allowed it only
+scale vertically.  However, with the introduction of both external clustering
+mechanisms (where a so-called proxy would distribute and balance requests across
+different Redis instances) and internal clustering support, Redis can now scale
+horizontally as well. Because of the variety of clustering solutions, and focus
+on Redis itself, external proxies are out of the scope of this evaluation.
+
+The integrated clustering solution of Redis is called \textit{Redis Cluster}.
+According to its documentation, \textit{"[it] is a distributed implementation
+of Redis"} \parencite{redis:clusterSpecification} and has three primary goals:
+
+\begin{itemize}
+    \item \textbf{High performance and linear scalability} up to 1000 nodes
+    \item \textbf{Write safety}
+    \item \textbf{Availability}
+\end{itemize}
+However, by specification, these criteria do not have to be guaranteed at all
+times -- altogether or even on their own. \parencite{redis:clusterSpecification}
+
+Nodes in a Redis Cluster setup are connected over \gls{tcp} bus connections.
+These connections (bus) are used to propagate information to all nodes in the
+cluster. Client requests to nodes in the cluster are not forwarded to the
+data-holding node. Instead, the client is redirected to the correct node by the
+use of \texttt{MOVED} or \texttt{ASK} error return codes.
+The data distribution is decided by the CRC-16 value of the string key. Each
+unique CRC-16 value is called a keyslot. All keyslots have one master and
+$n \geq 0$ slave nodes.
+\parencite{redis:clusterSpecification}
+
+\subsubsection*{Positioning in the CAP theorem}
+In the following, different criteria of Redis Cluster regarding typical usage
+guarantees in a distributed setup are evaluated:
+\begin{description}
+    \item \textbf{Consistency}:\\
```

Comment:
Maybe you might want to put the text about consistency in an `addmargin` environment or use something like `\leftskip`. This is more like personal taste, not a hard requirement.

Reply:
This also applies to other similar situations in the code. I am not exactly sure, however, as that would basically mean two margins.

Reply:
Yes, I get your point. However, after a quick offline discussion and a test-build, I think I'd prefer it the way it is.